garage/algos/ddqn.py

```diff
@@ -0,0 +1,339 @@
+        best_action = tensor.argmax(q_next, axis=1)
+        # Get the target quality value based on the best action
+        q_target_next = ll.get_output(
+            self._target_nn, deterministic=True, inputs=self._obs_next)
+        best_q_value = q_target_next[tensor.arange(q_target_next.shape[0]),
+                                     best_action]
+        target_value = (self._reward + self._gamma * self._done * best_q_value)
+
+        # Build loss expression
+        loss = lo.squared_error(old_q_value, target_value)
+        loss = lo.aggregate(loss, mode="mean")
+        # Build the update expression for the online network parameters
+        params = ll.get_all_params(self._online_nn, trainable=True)
+        updates = lu.sgd(loss, params, self._alpha)
+        # Build the functions to train the network and evaluate observations
+        self._learn = theano.function(
+            [
+                self._obs, self._obs_next, self._action, self._reward,
+                self._done
+            ],
+            loss,
+            updates=updates)
+        self._evaluate = theano.function([self._obs], q)
+
+    def get_action(self, obs):
+        """Get the best action based on the prediction of the Q function.
+
+        Parameters
+        ----------
+        obs : observation
+            Current observation of the environment.
+
+        """
+        qvals = self._evaluate(obs)
+        a = np.argmax(qvals)
+        return a
+
+    @overrides
+    def train(self):
+        """Training for all episodes."""
+        try:
+            for n in range(self._episodes):
+                self._train_per_episode(n)
+        finally:
+            self._env.close()
+
+    def _train_per_episode(self, n):
+        """Training for a single episode."""
+        observation = self._env.reset()
+        done = False
+        reward_episode = 0
+        while not done:
+            action = self._exploration_strategy.get_action(observation, self)
+            self._step_total += 1
+            next_observation, reward, done, _ = self._env.step(action)
+            self._memory.store(observation, action, next_observation, reward,
+                               not done)
+            if self._step_total >= self._exploration_steps:
+                batch = self._memory.get_batch(self._batch_size)
+                self._learn(batch.obs, batch.obs_next, batch.actions,
+                            batch.rewards, batch.not_done)
+                if not (self._step_total % self._tau):
+                    self._copy_to_target()
+            observation = next_observation
+            reward_episode += reward
+
+    def _copy_to_target(self):
+        """Copy weights from online network to target network."""
+        ll.set_all_param_values(self._target_nn,
+                                ll.get_all_param_values(self._online_nn))
+
+
+class MemoryBank:
+    """The memory bank is used for experience replay in DDQN."""
```

Comment:
I think this can be moved to garage.replay_buffer

Reply:
Also there is a replay_buffer in `garage.algos.ddpg` implementation already so if you can just move it out as a module and make use fo it, that would be great.